modules/platform/nuxeo-drive-server/nuxeo-drive-mongodb/src/main/java/org/nuxeo/drive/mongodb/MongoDBAuditChangeFinder.java

```diff
@@ -0,0 +1,179 @@
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.nuxeo.drive.service.SynchronizationRoots;
+import org.nuxeo.drive.service.impl.AuditChangeFinder;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.repository.RepositoryManager;
+import org.nuxeo.ecm.core.query.sql.model.OrderByExprs;
+import org.nuxeo.ecm.core.query.sql.model.Predicates;
+import org.nuxeo.ecm.core.query.sql.model.QueryBuilder;
+import org.nuxeo.ecm.platform.audit.api.AuditQueryBuilder;
+import org.nuxeo.ecm.platform.audit.api.AuditReader;
+import org.nuxeo.ecm.platform.audit.api.LogEntry;
+import org.nuxeo.mongodb.audit.MongoDBAuditBackend;
+import org.nuxeo.mongodb.audit.MongoDBAuditEntryReader;
+import org.nuxeo.runtime.api.Framework;
+
+import com.mongodb.client.MongoCollection;
+
+/**
+ * Override the JPA audit based change finder to execute query in BSON.
+ * <p>
+ * The structure of the query executed by the {@link AuditChangeFinder} is:
+ *
+ * <pre>
+ * from LogEntry log where log.repositoryId = :repositoryId
+ *
+ * + AND if ActiveRoots (activeRoots) NOT empty
+ *
+ * from LogEntry log where log.repositoryId = :repositoryId and (
+ * LIST_DOC_EVENTS_IDS_QUERY and ( ROOT_PATHS or COLECTIONS_PATHS) or
+ * (log.category = 'NuxeoDrive' and log.eventId != 'rootUnregistered') )
+ *
+ *
+ * if ActiveRoots EMPTY:
+ *
+ * from LogEntry log where log.repositoryId = :repositoryId and ((log.category =
+ * 'NuxeoDrive' and log.eventId != 'rootUnregistered'))
+ *
+ * + AND (log.id > :lowerBound and log.id <= :upperBound) + order by
+ * log.repositoryId asc, log.eventDate desc
+ * </pre>
+ *
+ * @since 11.3
+ */
+public class MongoDBAuditChangeFinder extends AuditChangeFinder {
+
+    private static final Logger log = LogManager.getLogger(MongoDBAuditChangeFinder.class);
+
+    @Override
+    public long getUpperBound() {
+        RepositoryManager repositoryManager = Framework.getService(RepositoryManager.class);
+        return getUpperBound(new HashSet<>(repositoryManager.getRepositoryNames()));
+    }
+
+    /**
+     * Returns the last available log id in the audit index considering events older than the last clustering
+     * invalidation date if clustering is enabled for at least one of the given repositories. This is to make sure the
+     * {@code DocumentModel} further fetched from the session using the audit entry doc id is fresh.
+     */
+    @Override
+    public long getUpperBound(Set<String> repositoryNames) {
+        long clusteringDelay = getClusteringDelay(repositoryNames);
+        MongoDBAuditBackend auditService = (MongoDBAuditBackend) Framework.getService(AuditReader.class);
+        QueryBuilder queryBuilder = new AuditQueryBuilder().predicate(
+                Predicates.in(LOG_REPOSITORY_ID, repositoryNames));
```

Comment:
What's the semantics of this predicate?
I think we can assume that any audit log entry has a repository id among the existing ones.

Reply:
This is my way of avoiding giving an AND() predicate with an empty array when clustering is disabled. Mongo would complain about it.
This is because the AuditQueryBuilder will init with an empty AND()

So I tell it to search all log repositories (if I guessEngineered correctly ?)
in(LOG_REPOSITORY_ID, repositoryNames));
means the field repositoryId's value is in the list of names repositoryNames.
So if I search all the logs that are in ["foo", "bar", "baz"]
if I get a log with {"repositoryId": "foo"} I will take it, same for other entries of the repositoryNames list "bar" and "baz"

Reply:
Oh! So it's a hack for a missing feature (no predicates) in `MongoDBAuditBackend`.
Might be interesting to have a new API in `MongoDBAuditBackend` such as:
```
public List<LogEntry> queryLogs(OrderByList orders, long offset, protected long limit)
```
@efge WDYT?

Reply:
I'm not sure I understand what the issue is here or why we need a new API. If clustering is disabled then there is no clustering delay stuff and nothing to query about repositories.

Reply:
If I pass an empty predicate here is what happens
```
com.mongodb.MongoQueryException: Query failed with error code 2 and error message '$and/$or/$nor must be a nonempty array' on server localhost:27017
	at com.mongodb.operation.FindOperation$1.call(FindOperation.java:735)
	at com.mongodb.operation.FindOperation$1.call(FindOperation.java:725)
	at com.mongodb.operation.OperationHelper.withReadConnectionSource(OperationHelper.java:463)
	at com.mongodb.operation.FindOperation.execute(FindOperation.java:725)
	at com.mongodb.operation.FindOperation.execute(FindOperation.java:89)
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:196)
	at com.mongodb.client.internal.MongoIterableImpl.execute(MongoIterableImpl.java:143)
	at com.mongodb.client.internal.MongoIterableImpl.iterator(MongoIterableImpl.java:92)
	at com.mongodb.client.internal.MongoIterableImpl.iterator(MongoIterableImpl.java:39)
	at java.base/java.lang.Iterable.spliterator(Iterable.java:101)
	at org.nuxeo.mongodb.audit.MongoDBAuditBackend.buildLogEntries(MongoDBAuditBackend.java:371)
	at org.nuxeo.mongodb.audit.MongoDBAuditBackend.queryLogs(MongoDBAuditBackend.java:186)
	at org.nuxeo.drive.mongodb.MongoDBAuditChangeFinder.getUpperBound(MongoDBAuditChangeFinder.java:119)
```

Reply:
This is because the [QueryBuilder](https://github.com/nuxeo/nuxeo/blob/master/modules/core/nuxeo-core-query/src/main/java/org/nuxeo/ecm/core/query/sql/model/QueryBuilder.java#L47) is initialized with and AND filter and, in [MongoDBAuditBackend](https://github.com/nuxeo/nuxeo/blob/master/modules/platform/nuxeo-platform-audit/nuxeo-platform-audit-mongodb/src/main/java/org/nuxeo/mongodb/audit/MongoDBAuditBackend.java#L178), we always apply the filters.

So, currently, there's no way to query the audit logs without any predicate through a `QueryBuilder`.

Reply:
> So, currently, there's no way to query the audit logs without any predicate through a `QueryBuilder`.

Yes because that's never been a use case. As a quick workaround if the `repositoryNames` is empty let's just replace the predicate with something like `gt(eventDate, new Date(0))` that should match everything and let us move forward, we'll refine it later.